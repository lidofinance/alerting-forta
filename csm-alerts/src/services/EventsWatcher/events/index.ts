export * from './assetRecoverer'
export * from './distributor'
export * from './accounting'
export * from './pausable'
export * from './proxies'
export * from './module'
export * from './oracle'
export * from './roles'
