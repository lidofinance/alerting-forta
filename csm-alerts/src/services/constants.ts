import { DeployedAddresses } from '../shared/types'

export const DEPLOYED_ADDRESSES: DeployedAddresses = {
    CS_MODULE: '0xdA7dE2ECdDfccC6c3AF10108Db212ACBBf9EA83F',
    CS_ACCOUNTING: '0x4d72BFF1BeaC69925F8Bd12526a39BAAb069e5Da',
    CS_FEE_DISTRIBUTOR: '0xD99CC66fEC647E68294C6477B40fC7E0F6F618D0',
    CS_FEE_ORACLE: '0x4D4074628678Bd302921c20573EEa1ed38DdF7FB',
    CS_VERIFIER: '0x3Dfc50f22aCA652a0a6F28a0F892ab62074b5583',
    CS_EARLY_ADOPTION: '0x3D5148ad93e2ae5DedD1f7A8B3C19E7F67F90c0E',
    CS_GATE_SEAL: '0x16Dbd4B85a448bE564f1742d5c8cCdD2bB3185D0',
    LIDO_STETH: '0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84',
    BURNER: '0xD15a672319Cf0352560eE76d9e89eAB0889046D3',
    HASH_CONSENSUS: '0x71093efF8D8599b5fA340D665Ad60fA7C80688e4',
    STAKING_ROUTER: '0xFdDf38947aFB03C621C71b06C9C70bce73f12999',
    MULTICALL3: '0xcA11bde05977b3631167028862bE2a173976CA11',
    PROVER: '0xDCBDb89D30E921C73B496e9d20a61d283A5D2401',
}

export const ALIASES: Record<keyof DeployedAddresses, string> = {
    CS_MODULE: 'CSM',
    CS_ACCOUNTING: 'CSAccounting',
    CS_FEE_DISTRIBUTOR: 'CSFeeDistributor',
    CS_FEE_ORACLE: 'CSFeeOracle',
    CS_VERIFIER: 'CSAccounting',
    CS_EARLY_ADOPTION: 'CSEarlyAdoption',
    CS_GATE_SEAL: 'CSM Gate Seal',
    LIDO_STETH: 'Lido',
    BURNER: 'Burner',
    HASH_CONSENSUS: 'CSM HashConsensus',
    STAKING_ROUTER: 'StakingRouter',
    MULTICALL3: 'Multicall3',
    PROVER: 'Prover Wallet',
}

export const ORACLE_MEMBERS: { [key: string]: string } = {
    '0x140Bd8FbDc884f48dA7cb1c09bE8A2fAdfea776E': 'Chorus One',
    '0x404335BcE530400a5814375E7Ec1FB55fAff3eA2': 'Staking Facilities',
    '0x946D3b081ed19173dC83Cd974fC69e1e760B7d78': 'Stakefish',
    '0x007DE4a5F7bc37E2F26c0cb2E8A95006EE9B89b5': 'P2P',
    '0xc79F702202E3A6B0B6310B537E786B9ACAA19BAf': 'Chainlayer',
    '0x61c91ECd902EB56e314bB2D5c5C07785444Ea1c8': 'bloXroute',
    '0x1Ca0fEC59b86F549e1F1184d97cb47794C8Af58d': 'Instadapp', // Revoked one.
    '0x73181107c8D9ED4ce0bbeF7A0b4ccf3320C41d12': 'Instadapp',
    '0xe57B3792aDCc5da47EF4fF588883F0ee0c9835C9': 'MatrixedLink',
    '0xA7410857ABbf75043d61ea54e07D57A6EB6EF186': 'Kyber Network',
}
