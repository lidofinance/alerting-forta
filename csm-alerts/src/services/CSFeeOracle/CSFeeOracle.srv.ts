import { elapsedTime } from '../../utils/time'
import { either as E } from 'fp-ts'
import { Logger } from 'winston'
import { BlockDto, EventOfNotice, TransactionDto, handleEventsOfNotice } from '../../entity/events'
import { Finding } from '../../generated/proto/alert_pb'
import { ethers, filterLog, getEthersProvider } from 'forta-agent'
import { ONE_DAY, ONE_MONTH, SECONDS_PER_SLOT } from '../../utils/constants'
import CS_FEE_ORACLE_ABI from '../../brief/abi/CSFeeOracle.json'
import HASH_CONSENSUS_ABI from '../../brief/abi/HashConsensus.json'
import { getLogsByChunks } from '../../utils/utils'
import {
  CS_FEE_ORACLE_REPORT_SUBMITTED_EVENT,
  HASH_CONSENSUS_REPORT_RECEIVED_EVENT,
} from '../../utils/events/cs_fee_oracle_events'
import BigNumber from 'bignumber.js'
export abstract class ICSFeeOracleClient {
  public abstract getBlockByNumber(blockNumber: number): Promise<E.Either<Error, BlockDto>>
}

interface MemberReport {
  refSlot: BigNumber
  report: string
  blockNumber: number
}

export class CSFeeOracleSrv {
  private readonly name = 'CSFeeOracleSrv'
  private readonly logger: Logger
  private readonly csFeeOracleClient: ICSFeeOracleClient

  private readonly hashConsensusEvents: EventOfNotice[]
  private readonly csFeeOracleEvents: EventOfNotice[]

  private readonly hashConsensusAddress: string
  private readonly csFeeOracleAddress: string

  private lastReportSubmitTimestamp: number = 0
  private lastReportOverdueAlertTimestamp: number = 0
  private membersAddresses: string[] = []
  private membersLastReport: Map<string, MemberReport> = new Map()

  constructor(
    logger: Logger,
    ethProvider: ICSFeeOracleClient,
    hashConsensusEvents: EventOfNotice[],
    csFeeOracleEvents: EventOfNotice[],
    hashConsensusAddress: string,
    csFeeOracleAddress: string,
  ) {
    this.logger = logger
    this.csFeeOracleClient = ethProvider

    this.hashConsensusAddress = hashConsensusAddress
    this.csFeeOracleAddress = csFeeOracleAddress

    this.hashConsensusEvents = hashConsensusEvents
    this.csFeeOracleEvents = csFeeOracleEvents
  }

  async getOracleMembers(blockNumber: number, hashConsensus: ethers.Contract): Promise<string[]> {
    const members = await hashConsensus.functions.getMembers({
      blockTag: blockNumber,
    })
    return members.addresses
  }

  public async initialize(currentBlock: number): Promise<Finding[] | null | Error> {
    const start = new Date().getTime()

    const currBlock = await this.csFeeOracleClient.getBlockByNumber(currentBlock)
    if (E.isLeft(currBlock)) {
      this.logger.error(elapsedTime(`Failed [${this.name}.initialize]`, start))
      return currBlock.left
    }

    const hashConsensus = new ethers.Contract(this.hashConsensusAddress, HASH_CONSENSUS_ABI, getEthersProvider())
    this.membersAddresses = await this.getOracleMembers(currentBlock, hashConsensus)
    const hashConsensusReportReceivedFilter = hashConsensus.filters.ReportReceived()
    const frameConfig = await hashConsensus.functions.getFrameConfig()
    const reportReceivedStartBlock = currentBlock - Math.ceil(frameConfig.epochsPerFrame * 32)
    const reportReceivedEvents = await getLogsByChunks(
      hashConsensus,
      hashConsensusReportReceivedFilter,
      reportReceivedStartBlock,
      currentBlock,
    )
    reportReceivedEvents.sort((a, b) => a.blockNumber - b.blockNumber)

    this.membersAddresses.forEach((member: string) => {
      const memberReports = reportReceivedEvents.filter((event) => {
        if (event.args) {
          return event.args.member == member
        }
      })
      if (memberReports.length > 0) {
        const lastReport = memberReports[memberReports.length - 1]
        if (lastReport.args) {
          this.membersLastReport.set(member, {
            refSlot: lastReport.args.refSlot,
            report: lastReport.args.report,
            blockNumber: lastReport.blockNumber,
          })
        }
      } else {
        this.membersLastReport.set(member, {
          refSlot: new BigNumber(0),
          report: '',
          blockNumber: 0,
        })
      }
    })

    const reportSubmits = await this.getReportSubmits(
      currentBlock - Math.ceil(ONE_MONTH / SECONDS_PER_SLOT),
      currentBlock,
    )

    if (reportSubmits.length > 0) {
      this.lastReportSubmitTimestamp = (await reportSubmits[reportSubmits.length - 1].getBlock()).timestamp
    }

    this.logger.info(elapsedTime(`[${this.name}.initialize]`, start))
    return null
  }

  public getName(): string {
    return this.name
  }

  async getReportSubmits(blockFrom: number, blockTo: number) {
    const csFeeOracle = new ethers.Contract(this.csFeeOracleAddress, CS_FEE_ORACLE_ABI, getEthersProvider())
    const oracleReportFilter = csFeeOracle.filters.ReportSubmitted()
    return await getLogsByChunks(csFeeOracle, oracleReportFilter, blockFrom, blockTo)
  }

  async handleBlock(blockDto: BlockDto): Promise<Finding[]> {
    const start = new Date().getTime()
    const findings: Finding[] = []

    const [reportOverdueFindings] = await Promise.all([this.handleReportOverdue(blockDto)])

    findings.push(...reportOverdueFindings)

    this.logger.info(elapsedTime(CSFeeOracleSrv.name + '.' + this.handleBlock.name, start))

    return findings
  }

  public async handleReportOverdue(blockDto: BlockDto): Promise<Finding[]> {
    const out: Finding[] = []

    const now = blockDto.timestamp

    const hashConsensus = new ethers.Contract(this.hashConsensusAddress, HASH_CONSENSUS_ABI, getEthersProvider())
    const frameConfig = await hashConsensus.functions.getFrameConfig()
    const MAX_REPORT_DELAY = frameConfig.epochsPerFrame * 32 * SECONDS_PER_SLOT

    const reportDelay = now - (this.lastReportSubmitTimestamp ? this.lastReportSubmitTimestamp : 0)

    const chainConfig = hashConsensus.getChainConfig()
    const genesisTimestamp = chainConfig.genesisTime
    const currentEpoch = Math.floor((now - genesisTimestamp) / 32 / SECONDS_PER_SLOT)
    if (currentEpoch < frameConfig.initialEpoch + frameConfig.epochsPerFrame) {
      // If we are still within the initial epochs, skip the overdue check
      return out
    }

    const timeSinceLastAlert = now - this.lastReportOverdueAlertTimestamp

    if (timeSinceLastAlert > ONE_DAY) {
      if (reportDelay > MAX_REPORT_DELAY) {
        const f: Finding = new Finding()
        f.setName(`ðŸ”´ CSFeeOracle: Report overdue`)
        f.setDescription(`Report is overdue by ${Math.floor(reportDelay / ONE_DAY)} days`)
        f.setAlertid('CS-FEE-ORACLE-REPORT-OVERDUE')
        f.setSeverity(Finding.Severity.HIGH)
        f.setType(Finding.FindingType.INFORMATION)
        f.setProtocol('ethereum')

        out.push(f)

        this.lastReportOverdueAlertTimestamp = now
      }
    }
    return out
  }

  public handleAlternativeHashReceived(txEvent: TransactionDto): Finding[] {
    const out: Finding[] = []

    const [event] = filterLog(txEvent.logs, HASH_CONSENSUS_REPORT_RECEIVED_EVENT, this.hashConsensusAddress)
    if (event && event.args) {
      const currentReportHashes = [...this.membersLastReport.values()]
        .filter((r) => r.refSlot.eq(event.args.refSlot))
        .map((r) => r.report)

      if (currentReportHashes.length > 0 && !currentReportHashes.includes(event.args.report)) {
        const f = new Finding()
        f.setName('ðŸ”´ HashConsensus: Another report variant appeared (alternative hash)')
        f.setDescription(
          `More than one distinct report hash received for slot ${event.args.refSlot}. Member: ${event.args.member}. Report: ${event.args.report}`,
        )
        f.setAlertid('HASH-CONSENSUS-REPORT-RECEIVED')
        f.setSeverity(Finding.Severity.HIGH)
        f.setType(Finding.FindingType.INFORMATION)
        f.setProtocol('ethereum')
        out.push(f)
      }

      this.membersLastReport.set(event.args.member, {
        refSlot: event.args.refSlot,
        report: event.args.report,
        blockNumber: txEvent.block.number,
      })
    }

    return out
  }

  public handleTransaction(txEvent: TransactionDto): Finding[] {
    const out: Finding[] = []

    const reportEvent = filterLog(txEvent.logs, CS_FEE_ORACLE_REPORT_SUBMITTED_EVENT, this.csFeeOracleAddress)
    if (reportEvent.length > 0) {
      this.lastReportSubmitTimestamp = txEvent.block.timestamp
    }

    const hashConsensusFindings = handleEventsOfNotice(txEvent, this.hashConsensusEvents)
    const csFeeOracleFindings = handleEventsOfNotice(txEvent, this.csFeeOracleEvents)
    const alternativeHashReceivedFindings = this.handleAlternativeHashReceived(txEvent)

    out.push(...hashConsensusFindings, ...csFeeOracleFindings, ...alternativeHashReceivedFindings)

    return out
  }
}
