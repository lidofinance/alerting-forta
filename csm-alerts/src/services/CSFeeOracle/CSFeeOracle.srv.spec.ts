import { DeploymentAddress, DeploymentAddresses } from '../../utils/constants.holesky'
import { expect } from '@jest/globals'
import { TransactionDto } from '../../entity/events'
import {
  CSModule__factory,
  CSAccounting__factory,
  CSFeeDistributor__factory,
  CSFeeOracle__factory,
} from '../../generated/typechain'
import { CSFeeOracleSrv, ICSFeeOracleClient } from './CSFeeOracle.srv'
import * as Winston from 'winston'
import { ETHProvider } from '../../clients/eth_provider'
import { ethers } from 'forta-agent'
import { getFortaConfig } from 'forta-agent/dist/sdk/utils'
import promClient from 'prom-client'
import { Metrics } from '../../utils/metrics/metrics'
import { getCSFeeOracleEvents, getHashConsensusEvents } from '../../utils/events/cs_fee_oracle_events'

const TEST_TIMEOUT = 120_000 // ms

describe('CSFeeOracle and HashConsensus events tests', () => {
  const chainId = 17000

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const address: DeploymentAddress = DeploymentAddresses

  const fortaEthersProvider = new ethers.providers.JsonRpcProvider(getFortaConfig().jsonRpcUrl, chainId)
  const csModuleRunner = CSModule__factory.connect(address.CS_MODULE_ADDRESS, fortaEthersProvider)
  const csAccountingRunner = CSAccounting__factory.connect(address.CS_ACCOUNTING_ADDRESS, fortaEthersProvider)
  const csFeeDistributorRunner = CSFeeDistributor__factory.connect(
    address.CS_FEE_DISTRIBUTOR_ADDRESS,
    fortaEthersProvider,
  )
  const csFeeOracleRunner = CSFeeOracle__factory.connect(address.CS_FEE_ORACLE_ADDRESS, fortaEthersProvider)

  const registry = new promClient.Registry()
  const m = new Metrics(registry, 'test_')

  const csFeeOracleClient: ICSFeeOracleClient = new ETHProvider(
    logger,
    m,
    fortaEthersProvider,
    csModuleRunner,
    csAccountingRunner,
    csFeeDistributorRunner,
    csFeeOracleRunner,
  )

  const csFeeOracleSrv = new CSFeeOracleSrv(
    logger,
    csFeeOracleClient,
    getHashConsensusEvents(address.HASH_CONSENSUS_ADDRESS),
    getCSFeeOracleEvents(address.CS_FEE_ORACLE_ADDRESS),
    address.HASH_CONSENSUS_ADDRESS,
    address.CS_FEE_ORACLE_ADDRESS,
  )

  test(
    'ðŸ”µ CSFeeOracle: Processing Started, Report Settled',
    async () => {
      const txHash = '0xf53cfcc9e576393b481a1c8ff4d28235703b6b5b62f9edb623d913b5d059f9c5'
      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = csFeeOracleSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(2)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”´ HashConsensus: FrameConfig Set',
    async () => {
      const txHash = '0xa6f4206ce30d66b378ab6e4ddef442ac4f29c95c5175fbb4a8944e6bec663724'
      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const result = csFeeOracleSrv.handleTransaction(transactionDto)

      expect(result).toMatchSnapshot()
      expect(result.length).toBe(1)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”´ HashConsensus: Member added, Quorum Set',
    async () => {
      const txHash = '0xdfcdbe0b9e795b2b83ad405c17b0c7326b00748cb3b11282a460c50b1f4588b0'

      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const results = csFeeOracleSrv.handleTransaction(transactionDto)

      expect(results).toMatchSnapshot()
      expect(results.length).toBe(2)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”´ HashConsensus: Member Removed, Quorum Set, Consensus Reached, Report Submitted',
    async () => {
      const txHash = '0xd22f8208b4bb2013a1113d68f9f19e3be13147c1f77ce811baa32ef082deed42'
      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const result = csFeeOracleSrv.handleTransaction(transactionDto)

      expect(result).toMatchSnapshot()
      expect(result.length).toBe(4)
    },
    TEST_TIMEOUT,
  )

  test(
    'Empty findings',
    async () => {
      const txHash = '0x74ff368ba6ea748e19a7f0fefd9d0f708078176e56799dbe97d46ae59782ff9d'
      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const result = csFeeOracleSrv.handleTransaction(transactionDto)

      expect(result.length).toBe(0)
    },
    TEST_TIMEOUT,
  )

  test(
    'ðŸ”´ CSFeeOracle: Consensus Hash Contract Set, Consensus Version Set, FeeDistributor Contract Set, Perf Leeway Set',
    async () => {
      const txHash = '0xdc5ed949e5b30a5ff6f325cd718ba5a52a32dc7719d3fe7aaf9661cc3da7e9a6'
      const trx = await fortaEthersProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        hash: trx.hash,
      }

      const result = csFeeOracleSrv.handleTransaction(transactionDto)

      expect(result).toMatchSnapshot()
      expect(result.length).toBe(4)
    },
    TEST_TIMEOUT,
  )
})
