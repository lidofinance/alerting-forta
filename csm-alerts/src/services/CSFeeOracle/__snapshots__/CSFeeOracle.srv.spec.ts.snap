// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSFeeOracle and HashConsensus events tests ðŸ”´ CSFeeOracle: Consensus Hash Contract Set, Consensus Version Set, FeeDistributor Contract Set, Perf Leeway Set 1`] = `
[
  {
    "array": [
      "ethereum",
      5,
      [],
      4,
      "CSFEE-ORACLE-CONSENSUS-HASH-CONTRACT-SET",
      "ðŸš¨ CSFeeOracle: Consensus hash contract set",
      "Consensus hash contract set to [0xbF38618Ea09B503c1dED867156A0ea276Ca1AE37](https://etherscan.io/address/0xbF38618Ea09B503c1dED867156A0ea276Ca1AE37), previous contract was [0x0000000000000000000000000000000000000000](https://etherscan.io/address/0x0000000000000000000000000000000000000000)",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0xbF38618Ea09B503c1dED867156A0ea276Ca1AE37,0x0000000000000000000000000000000000000000",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "CSFEE-ORACLE-CONSENSUS-VERSION-SET",
      "ðŸ”´ CSFeeOracle: Consensus version set",
      "Consensus version set to 1, previous version was 0",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "1,0",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "CSFEE-ORACLE-FEE-DISTRIBUTOR-CONTRACT-SET",
      "ðŸ”´ CSFeeOracle: New CSFeeDistributor set",
      "New CSFeeDistributor contract set to [0xD7ba648C8F72669C6aE649648B516ec03D07c8ED](https://etherscan.io/address/0xD7ba648C8F72669C6aE649648B516ec03D07c8ED)",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0xD7ba648C8F72669C6aE649648B516ec03D07c8ED",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "CSFEE-ORACLE-PERF-LEEWAY-SET",
      "ðŸ”´ CSFeeOracle: Performance leeway updated",
      "Performance leeway set to 500 basis points",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "500",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`CSFeeOracle and HashConsensus events tests ðŸ”´ HashConsensus: FrameConfig Set 1`] = `
[
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "HASH-CONSENSUS-FRAME-CONFIG-SET",
      "ðŸ”´ HashConsensus: Frame config set",
      "Frame configuration set. New initial epoch: 48038396020868878, Epochs per frame: 1575",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "48038396020868878,1575",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`CSFeeOracle and HashConsensus events tests ðŸ”´ HashConsensus: Member Removed, Quorum Set, Consensus Reached, Report Submitted 1`] = `
[
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "HASH-CONSENSUS-MEMBER-REMOVED",
      "ðŸ”´ HashConsensus: Member removed",
      "Member [0x4c75FA734a39f3a21C57e583c1c29942F021C6B7](https://etherscan.io/address/0x4c75FA734a39f3a21C57e583c1c29942F021C6B7) removed. Total members: 9, New quorum: 5",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0x4c75FA734a39f3a21C57e583c1c29942F021C6B7,9,5",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "HASH-CONSENSUS-QUORUM-SET",
      "ðŸ”´ HashConsensus: Quorum set",
      "Quorum set to 5. Total members: 9, Previous quorum: 6",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "5,9,6",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      1,
      [],
      4,
      "HASH-CONSENSUS-REACHED",
      "ðŸ”µ HashConsensus: Consensus reached, report received",
      "Consensus reached for slot 2068159. Report hash: 0x3f42b4c85f55a07ab3f38c6c910871fd654fe9dfed64c9ff69e8ad4921fd487b, Support: 5",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "2068159,0x3f42b4c85f55a07ab3f38c6c910871fd654fe9dfed64c9ff69e8ad4921fd487b,5",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      1,
      [],
      4,
      "CSFEE-ORACLE-REPORT-SUBMITTED",
      "ðŸ”µ CSFeeOracle: Report submitted",
      "Report submitted for slot 2068159. Hash: 0x3f42b4c85f55a07ab3f38c6c910871fd654fe9dfed64c9ff69e8ad4921fd487b, Processing deadline time: 1721325108",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "2068159,0x3f42b4c85f55a07ab3f38c6c910871fd654fe9dfed64c9ff69e8ad4921fd487b,1721325108",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`CSFeeOracle and HashConsensus events tests ðŸ”´ HashConsensus: Member added, Quorum Set 1`] = `
[
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "HASH-CONSENSUS-MEMBER-ADDED",
      "ðŸ”´ HashConsensus: Member added",
      "New member [0x4c75FA734a39f3a21C57e583c1c29942F021C6B7](https://etherscan.io/address/0x4c75FA734a39f3a21C57e583c1c29942F021C6B7) added. Total members: 10, New quorum: 6",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "0x4c75FA734a39f3a21C57e583c1c29942F021C6B7,10,6",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      4,
      [],
      4,
      "HASH-CONSENSUS-QUORUM-SET",
      "ðŸ”´ HashConsensus: Quorum set",
      "Quorum set to 6. Total members: 10, Previous quorum: 5",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "6,10,5",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;

exports[`CSFeeOracle and HashConsensus events tests ðŸ”µ CSFeeOracle: Processing Started, Report Settled 1`] = `
[
  {
    "array": [
      "ethereum",
      1,
      [],
      4,
      "CSFEE-ORACLE-PROCESSING-STARTED",
      "ðŸ”µ CSFeeOracle: Processing started",
      "Processing started for slot 2068159. Hash: 0x3f42b4c85f55a07ab3f38c6c910871fd654fe9dfed64c9ff69e8ad4921fd487b",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "2068159,0x3f42b4c85f55a07ab3f38c6c910871fd654fe9dfed64c9ff69e8ad4921fd487b",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
  {
    "array": [
      "ethereum",
      1,
      [],
      4,
      "CSFEE-ORACLE-REPORT-SETTLED",
      "ðŸ”µ CSFeeOracle: Report settled",
      "Report settled for slot 2068159. Distributed: 219882093527933929, Tree root: 0xfe5c8e3e617728bb0cd034313cc615f79ff7e93ff2e747b99d39b7e36a65b56a, Tree CID: QmV9AfsMa4zV3J1AjMhNuTQP4F18eV3uY5ihbVFcT5SkW6",
      ,
      ,
      [],
      ,
      [],
      [],
    ],
    "arrayIndexOffset_": -1,
    "convertedPrimitiveFields_": {},
    "messageId_": undefined,
    "pivot_": 1.7976931348623157e+308,
    "wrappers_": {
      "3": {
        "arrClean": false,
        "arr_": [],
        "map_": {
          "args": {
            "key": "args",
            "value": "2068159,219882093527933929,0xfe5c8e3e617728bb0cd034313cc615f79ff7e93ff2e747b99d39b7e36a65b56a,QmV9AfsMa4zV3J1AjMhNuTQP4F18eV3uY5ihbVFcT5SkW6",
            "valueWrapper": undefined,
          },
        },
        "valueCtor_": null,
      },
    },
  },
]
`;
